packages/flutter/lib/src/material/bottom_navigation.dart

```diff
@@ -0,0 +1,536 @@
+  /// The callback that is called when an action is tapped.
+  final BottomNavigationCallback onAction;
+
+  /// The type of the bottom navigation that effects its layout.
+  final BottomNavigationType type;
+
+  @override
+  _BottomNavigationState createState() => new _BottomNavigationState();
+}
+
+class _BottomNavigationState extends State<BottomNavigation> {
+  int currentIndex;
+  List<AnimationController> _controllers;
+  List<CurvedAnimation> _animations;
+  double _weight;
+  final Queue<_Circle> _circles = new Queue<_Circle>();
+  Color _backgroundColor;
+
+  static final Tween<double> _ratioTween = new Tween<double>(
+    begin: 1.0,
+    end: 1.5
+  );
+
+  @override
+  void initState() {
+    super.initState();
+    _controllers = new List<AnimationController>.generate(config.actions.length, (int index) {
+      return new AnimationController(
+        duration: const Duration(milliseconds: 200)
+      )..addListener(_handleTick);
+    });
+    _animations = new List<CurvedAnimation>.generate(config.actions.length, (int index) {
+      return new CurvedAnimation(
+        parent: _controllers[index],
+        curve: Curves.fastOutSlowIn,
+        reverseCurve: Curves.fastOutSlowIn.flipped
+      );
+    });
+
+    currentIndex = 0;
+    _controllers[currentIndex].value = 1.0;
+    if (config.onAction != null)
+      config.onAction(_animations[currentIndex], currentIndex);
+
+    _backgroundColor = config.actions[currentIndex].backgroundColor;
+  }
+
+  @override
+  void dispose() {
+    for (AnimationController controller in _controllers)
+      controller.dispose();
+    for (_Circle circle in _circles)
+      circle.dispose();
+    super.dispose();
+  }
+
+  void _handleTick() {
+    setState(() {
+      // Animation tick.
+    });
+  }
+
+  double get _maxWidth {
+    assert(config.type != null);
+    switch (config.type) {
+      case BottomNavigationType.fixed:
+        return config.actions.length * _kActiveMaxWidth;
+      case BottomNavigationType.shifting:
+        return _kActiveMaxWidth + (config.actions.length - 1) * _kInactiveMaxWidth;
+    }
+    return null;
+  }
+
+  bool _isAnimating(Animation<double> animation) {
+    return animation.status == AnimationStatus.forward ||
+           animation.status == AnimationStatus.reverse;
+  }
+
+  // Because of non-linear nature of the animations, the animations that are
+  // currently animating might not add up to the flex weight we are expecting.
+  // (1.5 + N - 1, since the max flex that the animating ones can have is 1.5)
+  // This causes jankiness in the animation when multiple actions are tapped.
+  // To solves this, we're always storing a weight that normalizes animating
+  // animations such that their resulting flex values will add up to the desired
+  // value.
+  void _computeWeight() {
+    final Iterable<Animation<double>> animating = _animations.where(
+      (Animation<double> animation) => _isAnimating(animation)
+    );
+
+    if (animating.length != 0) {
```

Comment:
i'm surprised the analyzer didn't complain; this should be "animating.isNotEmpty" (which will be quicker, since it'll abort at the first matching animation instead of counting all of them)